apps/design/design.md

```diff
@@ -0,0 +1,159 @@
+```
+
+Internals:
+```
+apps/
+  internal/
+    client/ - Shared package for common calls that Public and Confidential apps share
+    json/ - Our own json encoder/decoder for special needs
+    shared/ - Holds types that need to be in multiple packages and can't be moved into a single one due to import cycles
+    requests/ - The pacakge to communicate to services to get tokens
+```
+
+### Use of the Go special internal/ directory
+
+In Go, a directory called internal/ contains packages that should only be used by other packages
+rooted at the same location.
+
+This is documented here: https://golang.org/doc/go1.4#internalpackages
+
+For example, a package .../a/b/c/internal/d/e/f can be imported only by code in the directory tree rooted at .../a/b/c. It cannot be imported by code in .../a/b/g or in any other repository.
+
+We use this featurs quite liberally to make clear what is using an internal package.  For example:
+
+```
+apps/internal/base - Only can be used by packages defined at apps/
+apps/internal/base/internal/storage - Only can be use by package client
+```
+
+## Public API
+
+The public API will be encapsulated in apps/.  apps/ has 3 packages of interest to users:
+
+- public/ - This is what MSAL calls the Public Application Client (service client)
+- confidential/ - This is what MSAL calls the Confidential Application Client (service)
+- cache/ - This provides the interfaces that must be implemented to create peristent caches for any MSAL client
+
+## Internals
+
+In this section we will be talking about internal/.
+
+### JSON Handling
+
+JSON must be handled specially in our app. The basics are, if we receive fields that our
+structs do not contain, we cannot drop them.  We must send them back to the service.
+
+To handle that, we use our own custom json package that handles this.
+
+See the design at: [Design](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/internal/json/design.md)
+
+### Backend communication
+
+Communication to the backends is done via the requests/ package. requests.Token is the client
+for all communication.
+
+requests.Token communicates via REST calls that are encapsulated in the ops/ client.
+
+## Adding A Feature
+
+This is the general way to add a new feature to MSAL:
+
+- Add the REST calls to ops.REST
+- Add the higher level manipulations to requests.Token
+- Add your logic to the app/\<client\> and access the services via your requests.Token
+
+## Notable Differences To Other Clients
+
+### Confidential applications have a user ID
+
+The MSAL caching design is rather simple. These design decisions and the fact that multiple applications in different languages can share a cache mean it cannot be easily changed.
+
+The entire cache contents of a confidential.Client is read and written on 
+almost any action to and from an external cache. 
+
+It is not clear to a user that a confidential client should be per user to prevent scaling
```

Comment:
>per user  [](start = 63, length = 9)

See above and other comments I made regarding the cache elsewhere. we can take a follow up item to make changes to this. @patil.abhidnya I believe you are already tracking this, let's ensure this gets  captured in an appropriate work item

Reply:
I think someone will need to walk me through what you want changed in the cache.  This is mostly a simplification of what was provided in the original code.  Someone will need to walk me through the changes you want.

As for the integration tests, those are still with @abhidnya13 .

Reply:
Tracking this here : https://github.com/AzureAD/microsoft-authentication-library-for-go/pull/117#discussion_r553581029

Reply:
@henrik-me  does this need to go in this CL or can go in another with a new item tracker?  Cache behavior hasn't changed.